rust/lance/src/catalog/catalog_trait.rs

```diff
@@ -0,0 +1,276 @@
+use object_store::path::Path;
+use std::fmt::Debug;
+use std::sync::Arc;
+use url::Url;
+
+use crate::dataset::builder::DatasetBuilder;
+use crate::dataset::WriteParams;
+use crate::{Dataset, Result};
+
+/// Contains all the information that Lance needs to access a table
+pub struct TableReference {
+    /// Base URI of the table
+    pub uri: String,
+    /// Object store wrapper used to access the table
+    pub object_store: ObjectStore,
+    /// Commit handler used to handle new commits to the table
+    pub commit_handler: Arc<dyn CommitHandler>,
+    /// Parameters used to read / write to the object store
+    pub store_params: Option<ObjectStoreParams>,
+}
+
+/// Trait to be implemented by any catalog that Lance can use
+#[async_trait::async_trait]
+pub trait Catalog: std::fmt::Debug + Send + Sync {
+    /// Create a new table in the catalog
+    ///
+    /// Returns a table reference that can be used to read/write to the table
+    async fn create_table(&self, name: &str, schema: Arc<Schema>) -> Result<TableReference>;
+    /// Drop a table from the catalog
+    async fn drop_table(&self, name: &str) -> Result<()>;
+    /// Get a reference to a table in the catalog
+    ///
+    /// Returns a table reference that can be used to read/write to the table
+    async fn get_table(&self, name: &str) -> Result<TableReference>;
+    /// List all tables in the catalog
+    ///
+    /// The `start_after` parameter is an optional table name that, if provided, will
+    /// start the listing after the named table. If the named table is not found, the
+    /// listing will start after the table that would be named if it existed.
+    ///
+    /// The `limit` parameter is an optional limit on the number of tables returned.
+    async fn list_tables(
+        &self,
+        start_after: Option<String>,
+        limit: Option<u32>,
+    ) -> Result<Vec<String>>;
+}
+
+/// A simplified catalog that puts all tables in a single base directory
+///
+/// The object store's CAS primitives are used for commit handling.
+///
+/// This object store is simplistic but has zero dependencies (beyond an object store
+/// of some kind)
+#[derive(Debug)]
+pub struct ListingCatalog {
+    base_path: Path,
+    object_store: ObjectStore,
+    commit_handler: Arc<dyn CommitHandler>,
+}
+
+const LANCE_EXTENSION: &str = "lance";
+
+fn format_table_url_or_path(is_url: bool, scheme: &str, base_path: &str, name: &str) -> String {
+    if is_url {
+        format!("{}:///{}/{}.{}", scheme, base_path, name, LANCE_EXTENSION)
+    } else {
+        format!("{}/{}.{}", base_path, name, LANCE_EXTENSION)
+    }
+}
+
+impl ListingCatalog {
+    pub fn new(base_path: Path, object_store: ObjectStore) -> Self {
+        Self {
+            base_path,
+            object_store: object_store.clone(),
+            commit_handler: Arc::new(ListingCommitHandler { object_store }),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Catalog for ListingCatalog {
```

Comment:
@westonpace I implemented a simple version for discussion. Please take a look when you have time.

Reply:
@westonpace Any feedback here? Please ignore the catalog trait's location currently. I define it here to avoid the prototype implementation having a big span between two packages.